x/sudo/sudo.go

```diff
@@ -0,0 +1,236 @@
+			return nil, sdkerrors.Wrap(sdkerrors.ErrUnknownRequest, errMsg)
+		}
+	}
+}
+
+// Ensure the interface is properly implemented at compile time
+var _ pb.MsgServer = Keeper{}
+
+func (k Keeper) EditSudoers(
+	goCtx context.Context, msg *pb.MsgEditSudoers,
+) (*pb.MsgEditSudoersResponse, error) {
+	switch msg.Action {
+	case ROOT_ACTION.AddContracts:
+		return k.AddContracts(goCtx, msg)
+	case ROOT_ACTION.RemoveContracts:
+		return k.RemoveContracts(goCtx, msg)
+	default:
+		return nil, msg.ValidateBasic()
+	}
+}
+
+func (k Keeper) SenderHasPermission(sender string, root string) error {
+	if sender != root {
+		return fmt.Errorf(`message must be sent by root user. root: "%s", sender: "%s"`,
+			root, sender,
+		)
+	}
+	return nil
+}
+
+// ————————————————————————————————————————————————————————————————————————————
+// Encoder for the Sudoers type
+// ————————————————————————————————————————————————————————————————————————————
+
+func SudoersValueEncoder(cdc codec.BinaryCodec) collections.ValueEncoder[pb.Sudoers] {
+	return collections.ProtoValueEncoder[pb.Sudoers](cdc)
+}
+
+type PbSudoers = pb.Sudoers
+
+type Sudoers struct {
+	Root      string          `json:"root"`
+	Contracts set.Set[string] `json:"contracts"`
+}
+
+func (sudo Sudoers) FromPbSudoers(pbSudoers pb.Sudoers) Sudoers {
+	return Sudoers{
+		Root:      pbSudoers.Root,
+		Contracts: set.New[string](pbSudoers.Contracts...),
+	}
+}
+
+func (sudo Sudoers) String() string {
+	jsonBz, _ := sudo.MarshalJSON()
+	return string(jsonBz)
+}
+
+func (sudo Sudoers) MarshalJSON() ([]byte, error) {
+	return json.Marshal(sudo)
+}
+
+func SudoersToPb(sudo Sudoers) pb.Sudoers {
+	return pb.Sudoers{
+		Root:      sudo.Root,
+		Contracts: sudo.Contracts.ToSlice(),
+	}
+}
+
+// ————————————————————————————————————————————————————————————————————————————
+// AddContracts
+// ————————————————————————————————————————————————————————————————————————————
+
+// Sudoers.AddContracts adds contract addresses to the sudoer set.
+func (sudo *Sudoers) AddContracts(
+	contracts []string,
+) (out set.Set[string], err error) {
+	for _, contractStr := range contracts {
+		contract, err := sdk.AccAddressFromBech32(contractStr)
+		if err != nil {
+			return out, err
+		}
+		sudo.Contracts.Add(contract.String())
+	}
+	return sudo.Contracts, err
+}
+
+// Keeper.AddContracts executes a MsgEditSudoers message with action type
+// "add_contracts". This adds contract addresses to the sudoer set.
+func (k Keeper) AddContracts(
+	goCtx context.Context, msg *pb.MsgEditSudoers,
+) (msgResp *pb.MsgEditSudoersResponse, err error) {
+	if msg.Action != ROOT_ACTION.AddContracts {
+		err = fmt.Errorf("invalid action type %s for msg add contracts", msg.Action)
+		return
+	}
+
+	err = msg.ValidateBasic()
+	if err != nil {
+		return
+	}
+
+	// Read state
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	pbSudoersBefore, err := k.Sudoers.Get(ctx)
+	if err != nil {
+		return
+	}
+	sudoersBefore := Sudoers{}.FromPbSudoers(pbSudoersBefore)
```

Comment:
you can go wild and create a proto encoder for the Set type so you dot have to use this converters.